### PR TITLE
[WIP] fix #13404 poEvalCommand now works on windows

### LINKED ISSUE #13404: poEvalCommand works on posix but (silently) not on windows
poEvalCommand works on posix but (silently, ie no error) not on windows.
yes, poEvalCommand has security concerns, but doesn't mean it should work on posix but not on windows.

### Example
```nim
import osproc
echo execCmdEx("echo ok1 && echo ok2",options={poStdErrToStdOut, poUsePath, poEvalCommand})
echo execCmdEx("sh -c 'echo ok1 && echo ok2'")
```

### Current Output
```
(output: "ok1 && echo ok2\n", exitCode: 0)
(output: "ok1\nok2\n", exitCode: 0)
```

### Expected Output
```
(output: "ok1\nok2\n", exitCode: 0)
(output: "ok1\nok2\n", exitCode: 0)
```

### Possible Solution

* adapt this code snippet to windows implementation:
```nim
    if poEvalCommand in options:
      const useShPath {.strdefine.} =
        when not defined(android): "/bin/sh"
        else: "/system/bin/sh"
      data.sysCommand = useShPath
      sysArgsRaw = @[data.sysCommand, "-c", command]
      assert args.len == 0, "`args` has to be empty when using poEvalCommand."
```

* related (but not same): https://github.com/nim-lang/Nim/issues/8913

### PR DESCRIPTION
**Pull request by timotheecour**
* fix #13404 poEvalCommand now works on windows
* added runnableExample
* `execCmdEx("echo ok1 && echo ok2")` now works on posix and windows

### CODE REVIEW

File: lib/pure/osproc.nim [hunk 1]
```diff
@@ -70,6 +70,10 @@ const poDemon* {.deprecated.} = poDaemon ## Nim versions before 0.20
                                          ## Now `ProcessOption` uses the correct spelling ("daemon"),
                                          ## and this is needed just for backward compatibility.
 
+const useShPath {.strdefine.} =
+  when defined(windows): "sh"
```
Comment (Araq) [inline lib/pure/osproc.nim]: "sh" is not a common Windows tool.

**Comment by Araq** [comment]
Completely unacceptable approach, so now ``poEvalCommand`` for Windows depends on some Unix emulation layer?

**Comment by timotheecour** [comment]
* then what's `sh` best equivalent on windows, such that it would allow for example running 2 commands consecutively, eg:

 `execCmdEx("git checkout hotfix && nimble test")` ?

(not looking for a workaround like `doAssert execCmdEx("git checkout hotfix") == 0; doAssert execCmdEx("nimble test") == 0;` )

* running via sh / bash seems useful though, and would help cross platform code.

I could either add a param `sh` to `execProcess` + friends, eg:
`proc execProcess(command: string, ..., sh = "")` 
(where sh becomes `useShPath` if empty, and `useShPath` defaults to empty on windows)

or I could add a windows specific flag `poEvalCommandViaSh` (but that seems less flexible, eg not overrideable as bash / sh)

The logic in this PR is non-trivial (see `quoteShellWindowsForSh` which is different from `quoteShellWindows`) so implementing it in properly in stdlib once instead of letting the user stumble upon it each time makes sense

**Comment by Clyybber** [comment]
Use cmd ?

**Comment by timotheecour** [comment]
> Use cmd ?

@Clyybber can you please provide a full example? how would I run 2 commands using `cmd.exe`, eg:
```nim
import osproc
echo execProcess("sh -c 'echo ok1 && echo ok2'") # ok, prints: ok1\nok2\n (on my windows machine)
echo execProcess("cmd -c 'echo ok1 && echo ok2'") # this hangs
```
for a concrete example, `git checkout hotfix && nimble test`

**Comment by Clyybber** [comment]
That is not the problem of poEvalCommand. It should just pass it on to the shell.
In this case use `cmd /c "..."`

File: lib/pure/os.nim [hunk 2]
```diff
@@ -965,6 +965,37 @@ proc quoteShellWindows*(s: string): string {.noSideEffect, rtl, extern: "nosp$1"
   if needQuote:
     result.add("\"")
 
+when defined(windows):
+  proc quoteShellWindowsForSh*(s: string): string {.noSideEffect, rtl, extern: "nosp$1".} =
```
Comment (Clyybber) [inline lib/pure/os.nim]: Remove this dead code please

Comment (timotheecour) [reply lib/pure/os.nim]: I've removed it to unblock this PR but it is useful in case user wants to use non-standard sh or bash instead of standard (but fraught with limitations) cmd.exe

**Comment by timotheecour** [comment]
> Completely unacceptable approach, so now poEvalCommand for Windows depends on some Unix emulation layer?

fixed; `cmd /c echo ok1 && echo ok2` does works, so I've updated PR to use `cmd /c` instead of `sh -c` on windows.
so now, `execCmdEx("echo ok1 && echo ok2")` does work cross-platform (and doesn't rely on non-standard tools)

File: lib/pure/osproc.nim [hunk 3]
```diff
@@ -70,6 +70,10 @@ const poDemon* {.deprecated.} = poDaemon ## Nim versions before 0.20
                                          ## Now `ProcessOption` uses the correct spelling ("daemon"),
                                          ## and this is needed just for backward compatibility.
 
+const useShPath {.strdefine.} =
+  when defined(windows): "cmd"
+  elif not defined(android): "/bin/sh"
+  else: "/system/bin/sh"
```
Comment (Clyybber) [inline lib/pure/osproc.nim]: Nit, but please do
```
elif defined(android): "/system/bin/sh"
else: "/bin/sh"
```
instead

Comment (timotheecour) [reply lib/pure/osproc.nim]: was pre-existing, but good point, fixed

File: lib/pure/osproc.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/osproc.nim]: fixed => now using `cmd /c`

**Review by Clyybber — APPROVED** [review]

**Comment by Araq** [comment]
> (not looking for a workaround like doAssert execCmdEx("git checkout hotfix") == 0; doAssert execCmdEx("nimble test") == 0; )

That's not a "workaround", it's the only way to do this properly. Otherwise you're in shell quoting hell with all its non-portability and security nightmares.

**Comment by timotheecour** [comment]
> That's not a "workaround", it's the only way to do this properly. Otherwise you're in shell quoting hell with all its non-portability and security nightmares.

I know, but sometimes it's inconvenient to modify third party code that was written with single command in mind just to support chaining multiple commands. Example: 
in important_packages.nim, if you want to apply a hotfix branch to a package that's broken (and the author is unresponsive or would be inconvenient to merge the hotfix right away), all you'd need (after this PR) is:
```
pkg "foo", "nimble test", url = "https://github.com/realauthor/foo"
=>
pkg "foo", "git checkout hotfix && nimble test", url = "https://github.com/hotfixauthor/foo"
```
and you don't have to rewrite the plubming to support seq[string] in important_packages.nim. (that was just an example; my point remains even if we do end up supporting seq[string] in this particular case)

in any case, we should strive for minimizing cross-platform differences, and now `execCmdEx("git checkout hotfix && nimble test")` works now cross platform

File: testament/categories.nim [hunk 4]
```diff
@@ -101,7 +101,7 @@ proc flagTests(r: var TResults, cat: Category, options: string) =
   testSpec r, makeTest(filename, genopts, cat)
 
   when defined(windows):
-    testExec r, makeTest(filename, " cmd /c cd " & nimcache &
+    testExec r, makeTest(filename, " cd " & nimcache &
```
Comment (Clyybber) [inline testament/categories.nim:104]: Out of curiousity, why is this needed?

Comment (timotheecour) [reply testament/categories.nim:104]: in this particular case `cmd /c` would already be added by this PR's logic so I'm removing it from here; I am trying to fix CI failures; help welcome btw
* see https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/2606/logs/68
* and https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/2606/logs/72

Comment (Clyybber) [reply testament/categories.nim:104]: @timotheecour Sure, hop on gitter/irc?
